tests/db_functions/json/test_json_array.py

```diff
@@ -0,0 +1,122 @@
+from django.db.models.functions import Cast, JSONArray, Lower
+from django.test import TestCase
+from django.test.testcases import skipIfDBFeature, skipUnlessDBFeature
+from django.utils import timezone
+
+from ..models import Article, Author
+
+
+@skipUnlessDBFeature("has_json_object_function")
+class JSONArrayTests(TestCase):
+    @classmethod
+    def setUpTestData(cls):
+        Author.objects.bulk_create(
+            [
+                Author(name="Ivan Ivanov", alias="iivanov"),
+                Author(name="Bertha Berthy", alias="bberthy"),
+            ]
+        )
+
+    def test_empty(self):
+        obj = Author.objects.annotate(json_array=JSONArray()).first()
+        self.assertEqual(obj.json_array, [])
+
+    def test_basic(self):
+        obj = Author.objects.annotate(
+            json_array=JSONArray(Value("name"), F("name"))
+        ).first()
+        self.assertEqual(obj.json_array, ["name", "Ivan Ivanov"])
+
+    def test_expressions(self):
+        obj = Author.objects.annotate(
+            json_array=JSONArray(
+                Lower("name"),
+                F("alias"),
+                F("goes_by"),
+                Value(30000.15),
+                F("age") * 2,
+            )
+        ).first()
+        self.assertEqual(
+            obj.json_array,
+            [
+                "ivan ivanov",
+                "iivanov",
+                None,
+                30000.15,
+                60,
+            ],
+        )
+
+    def test_nested_json_array(self):
+        obj = Author.objects.annotate(
+            json_array=JSONArray(
+                F("name"),
+                JSONArray(F("alias"), F("age")),
+            )
+        ).first()
+        self.assertEqual(
+            obj.json_array,
+            [
+                "Ivan Ivanov",
+                ["iivanov", 30],
+            ],
+        )
+
+    def test_nested_empty_json_array(self):
+        obj = Author.objects.annotate(
+            json_array=JSONArray(
+                F("name"),
+                JSONArray(),
+            )
+        ).first()
+        self.assertEqual(
+            obj.json_array,
+            [
+                "Ivan Ivanov",
+                [],
+            ],
+        )
+
+    def test_textfield(self):
+        Article.objects.create(
+            title="The Title",
+            text="x" * 4000,
+            written=timezone.now(),
+        )
+        obj = Article.objects.annotate(json_array=JSONArray(F("text"))).first()
+        self.assertEqual(obj.json_array, ["x" * 4000])
+
+    @unittest.skipUnless(connection.vendor == "postgresql", "PostgreSQL specific tests")
+    def test_explicit_cast(self):
+        obj = Author.objects.annotate(json_array=JSONArray(Cast("age", CharField())))
+
+        self.assertTrue(
+            bool(re.search(r"::varchar", str(obj.query))),
+        )
+        self.assertFalse(
+            bool(re.search(r"::varchar\)?::varchar", str(obj.query))),
+        )
```

Comment:
`re.search` seems to be unnecessary. I'd assert actually executed query:
```suggestion
        with self.assertNumQueries(1) as ctx:
            self.assertSequenceEqual(
                Author.objects.annotate(
                    json_array=JSONArray(Cast("age", CharField()))
                ).values("json_array"),
                [   
                    {"json_array": ["30"]},
                    {"json_array": ["30"]},
                ],
            )
        sql = ctx.captured_queries[0]["sql"]
        self.assertIn("::varchar", sql)
        self.assertNotIn("::varchar)::varchar", sql)
```

Reply:
Great! I wasn't familiar with those assertions, nor the captured query pattern, so this is definitely more in line what I would have authored had I known.